File: arrays_and_strings/number_of_ways_to_split_array.cpp
#include <gtest/gtest.h>

#include <vector>

//! @brief Get number of ways to split array such that LHS sum >= RHS sum
//! @param[in] nums Vector of integers
//! @return Number of ways to split input
static int waysToSplitArray1(std::vector<int> nums)
{
    //! @details Time complexity O(N) since prefix sums allow LHS/RHS sums to be
    //!          evaluated in O(1)
    //!          Space complexity O(N) for prefix sum vector

    const auto nums_size = static_cast<int>(nums.size());

    std::vector<long> prefix = {nums[0]};
    for (int i = 1; i < nums_size; ++i)
    {
        prefix.push_back(nums[i] + prefix.back());
    }

    int ans {};
    for (int i = 0; i < nums_size - 1; ++i)
    {
        long leftSection = prefix[i];
        long rightSection = prefix.back() - prefix[i];
        if (leftSection >= rightSection)
        {
            ++ans;
        }
    }

    return ans;

} // static int waysToSplitArray1( ...

TEST(WaysToSplitArrayTest, SampleTest)
{
    EXPECT_EQ(2, waysToSplitArray1({10, 4, -8, 7}));
}

File: arrays_and_strings/number_of_ways_to_split_array.py
def waysToSplitArray1(nums: List[int]) -> int:
    prefix = [nums[0]]
    for i in range(1, len(nums)):
        prefix.append(nums[i] + prefix[-1])

    ans = 0
    for i in range(len(nums) - 1):
        left_section = prefix[i]
        right_section = prefix[-1] - prefix[i]
        if left_section >= right_section:
            ans += 1
    
    return ans

def waysToSplitArray2(nums: List[int]) -> int:
    ans = left_section = 0
    total = sum(nums)

    for i in range(len(nums) - 1):
        left_section += nums[i]
        right_section = total - left_section
        if left_section >= right_section:
            ans += 1

    return ans
